File: nodejs/src/Graph/NumberOfIslands/NumberOfIslands.ts
class NumberOfIslands {
  grid: string[][]

  constructor(grid: string[][]) {
    this.grid = grid
  }

  exec(): number {
    const ROWS: number = this.grid.length;
    const COLS: number = this.grid[0].length;
    const visited: Set<string> = new Set<string>();

    // if we can find an unexplored island from (i, j), returning true
    const explore = (i: number, j: number): boolean => {
      if(i < 0 || i >= ROWS || j < 0 || j >= COLS || this.grid[i][j] === '0') return false;
      const key: string = `${i},${j}`
      if(visited.has(key)) return false;

      visited.add(key)
      explore(i, j - 1)
      explore(i - 1, j)
      explore(i, j + 1)
      explore(i + 1, j)

      return true;
    }

    let res: number = 0;
    for(let i = 0; i < ROWS; i++){
      for(let j = 0; j < COLS; j++){
        if(explore(i, j) === true) res++
      }
    }

    return res;
  }
}

export default NumberOfIslands

File: python/src/graph/number-of-islands/number-of-islands.py
from typing import List


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        # get boundary values for 2-D problems first
        ROWS: int = len(grid)
        COLS: int = len(grid[0])

        # stores visited coordinate
        visited = set()

        # final result count
        count = 0

        def dfs(row: int, col: int) -> bool:
            """return true if we can find an island starting from (row, col)

            The side effect is that we will mark the visited location
            """

            # fail conditions
            if row < 0 or row >= ROWS or col < 0 or col >= COLS: # out of boudary
                return False
            if grid[row][col] == '0':   # find water
                return False
            if (row, col) in visited:   # visited already
                return False

            # mark (row, col) as visited
            visited.add((row, col))

            # visit neighbours and mark them
            dfs(row, col - 1)
            dfs(row - 1, col)
            dfs(row, col + 1)
            dfs(row + 1, col)

            return True

        # start from each location and check results
        for i in range(ROWS):
            for j in range(COLS):
                if dfs(i, j):
                    count += 1

        return count


if __name__ == '__main__':
    grid1 = [
      ["1","1","1","1","0"],
      ["1","1","0","1","0"],
      ["1","1","0","0","0"],
      ["0","0","0","0","0"]
    ]
    solution1: Solution = Solution()
    assert solution1.numIslands(grid1) == 1

    grid2 = [
      ["1","1","0","0","0"],
      ["1","1","0","0","0"],
      ["0","0","1","0","0"],
      ["0","0","0","1","1"]
    ]
    solution2: Solution = Solution()
    assert solution2.numIslands(grid2) == 3

